### ISSUE TITLE
Conv2D is 20x slower using mixed_float16 in training process

### ISSUE THREAD

**Issue by leondgarse** [tensorflow/tensorflow#48127]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): `v2.4.0-49-g85c8b2a817f 2.4.1`
- Python version: `3.7.6`
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: `CUDA Version: 11.2`
- GPU model and memory: `Quadro RTX 8000, compute capability 7.5` and also `GeForce RTX 2080 Ti, compute capability 7.5`

**Describe the current behavior**
- `Conv2D` in training process is `~20x` slower if set `mixed_precission` `mixed_float16`. In my test, this happens in some conditions. In a short result, it's:
  | input_shape | batch_size | use_fp16 | first epoch (ms/step) | second epoch (ms/step) |
  | ----------- | ---------- | -------- | --------------------- | ---------------------- |
  | 112         | 512        | False    | 83                    | 67                     |
  | 112         | 512        | True     | 1969                  | 1857                   |

  Seems like `mixed_float16` is `~20x` times slower than `float32`.

**Describe the expected behavior**
- Expect `mixed_float16` performs better than `float32`.

**Standalone code to reproduce the issue**
- I just used a minimal python script to test this scenario, which seems easier for me:
  ```py
  import tensorflow as tf
  from tensorflow import keras

  gpus = tf.config.experimental.list_physical_devices("GPU")
  for gpu in gpus:
      tf.config.experimental.set_memory_growth(gpu, True)

  def test_resnet(input_shape=None, classes=1000):
      img_input = keras.layers.Input(shape=input_shape)
      nn = img_input

      # nn = keras.layers.ZeroPadding2D(padding=((2, 2), (2, 2)), name='conv1_pad')(nn) # NOT helping
      # nn = keras.layers.ZeroPadding2D(padding=((3, 3), (3, 3)), name='conv1_pad')(nn) # NOT helping
      nn = keras.layers.Conv2D(64, 7, strides=2, use_bias=False, padding='VALID', name="conv1_conv")(nn)
      # nn = keras.layers.Conv2D(64, 7, strides=2, use_bias=False, padding='VALID', name="conv1_conv", data_format="channels_first")(nn)

      nn = keras.layers.Flatten()(nn)
      nn = keras.layers.Dense(classes, name="predictions")(nn)
      nn = keras.layers.Activation('softmax', dtype='float32')(nn)
      return keras.models.Model(img_input, nn)

  def load_cifar10(batch_size=1024, image_shape=(32, 32)):
      import tensorflow_datasets as tfds
      AUTOTUNE = tf.data.experimental.AUTOTUNE

      if image_shape[:2] == (32, 32):
          preprocessing = lambda data: (tf.cast(data["image"], tf.float32) / 255.0, data["label"])
      else:
          preprocessing = lambda data: (tf.image.resize(data["image"], image_shape[:2]) / 255.0, data["label"])
          # preprocessing = lambda data: (tf.transpose(tf.image.resize(data["image"], image_shape[1:]) / 255.0, [2, 0, 1]), data["label"])
      dataset = tfds.load("cifar10", split="train").map(preprocessing, num_parallel_calls=AUTOTUNE)
      dataset = dataset.cache().batch(batch_size).prefetch(buffer_size=AUTOTUNE)
      return dataset

  def run_test(input_shape=(32, 32, 3), batch_size=512, use_fp16=False, epochs=2):
      if use_fp16:
          keras.mixed_precision.set_global_policy('mixed_float16')
          # tf.config.optimizer.set_experimental_options({"auto_mixed_precision": True})

      dataset = load_cifar10(batch_size=batch_size, image_shape=input_shape)

      # model = keras.applications.ResNet50(input_shape=input_shape, classes=10, weights=None)
      # model = keras.models.Model(model.inputs[0], keras.layers.Activation("linear", dtype="float32")(model.outputs[0]))
      model = test_resnet(classes=10, input_shape=input_shape)

      # optimizer = keras.mixed_precision.LossScaleOptimizer(keras.optimizers.Adam())
      optimizer = keras.optimizers.Adam()
      model.compile(loss='sparse_categorical_crossentropy', optimizer=optimizer)
      history = model.fit(dataset, epochs=epochs)

  if __name__ == "__main__":
      import sys
      import argparse

      parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
      parser.add_argument("-b", "--batch_size", type=int, default=512, help="Batch size")
      parser.add_argument("-i", "--input_shape", type=int, default=32, help="Input shape")
      parser.add_argument("-f", "--use_fp16", action="store_true", help="Use fp16")
      parser.add_argument("-e", "--epochs", type=int, default=2, help="Epochs")

      args = parser.parse_known_args(sys.argv[1:])[0]
      run_test((args.input_shape, args.input_shape, 3), args.batch_size, args.use_fp16, args.epochs)
      # keras.backend.set_image_data_format('channels_first') # NOT helping
      # run_test((3, args.input_shape, args.input_shape), args.batch_size, args.use_fp16, args.epochs)
  ```

**Other info / logs**
- That `Conv2D` layer in `test_resnet` is actually the head layer in our `keras.applications.ResNet50`, and I found it's this layer slowing down the model training speed in `mixed_float16`.
- By changing the `input_shape` and `batch_size`, it seems `mixed_float16` working good in `input_shape == 224, batch_size == 512` or `input_shape == 112, batch_size == 256`.
  | input_shape | batch_size | use_fp16 | first epoch (ms/step) | second epoch (ms/step) |
  | ----------- | ---------- | -------- | --------------------- | ---------------------- |
  | 112         | 512        | False    | 83                    | 67                     |
  | 112         | 512        | True     | 1969                  | 1857                   |
  | 128         | 512        | False    | 96                    | 85                     |
  | 128         | 512        | True     | 96                    | 97                     |
  | 112         | 256        | False    | 43                    | 38                     |
  | 112         | 256        | True     | 46                    | 41                     |
  | 224         | 512        | False    | 233                   | 212                    |
  | 224         | 512        | True     | 228                   | 210                    |
  | 32          | 512        | False    | 8                     | 5                      |
  | 32          | 512        | True     | 38                    | 37                     |

- In some further tests in my environment:
  - First set `input_shape == 112`, and use different `batch_size`, seems this slowing happens if `batch_size` **>** `384`.
  - Then set `batch_size == 512`, and vary `input_shape`, seems it's happening if `input_shape` **<** `121`.

  | input_shape | batch_size | use_fp16 | first epoch (ms/step) | second epoch (ms/step) |
  | ----------- | ---------- | -------- | --------------------- | ---------------------- |
  | 112         | 384        | True     | 63                    | 45                     |
  | 112         | 385        | True     | 1407                  | 1353                   |
  | 120         | 512        | True     | 2306                  | 2183                   |
  | 121         | 512        | True     | 90                    | 84                     |
  | 122         | 512        | True     | 88                    | 66                     |

  Also, if we add a `nn = keras.layers.ZeroPadding2D(padding=((3, 3), (3, 3)), name='conv1_pad')(nn)` layer before `Conv2D`, it will be `input_shape < 115 == 121 - 6`.
- Then I set `input_shape == 112, batch_size == 512`, and change parameters in `Conv2D`, seems `padding="SAME"` working good in `mixed_float16`:

  | kernel_size | padding | use_fp16 | first epoch (ms/step) | second epoch (ms/step) |
  | ----------- | ------- | -------- | --------------------- | ---------------------- |
  | 7           | SAME    | True     | 77                    | 60                     |
  | 7           | VALID   | True     | 1969                  | 1857                   |
  | 5           | SAME    | True     | 1132                  | 1061                   |
  | 5           | VALID   | True     | 953                   | 948                    |
  | 3           | SAME    | True     | 351                   | 342                    |
  | 3           | VALID   | True     | 339                   | 330                    |

- In `ResNet50` case, using `padding="VALID"` is `~4times slower` in `mixed_float16` training, but if we set `padding="SAME"` only in the first `Conv2D` layer, it's faster:

  | padding | float16 | first epoch (ms/step) | second epoch (ms/step) |
  | ------- | ------- | --------------------- | ---------------------- |
  | VALID   | False   | 608                   | 570                    |
  | VALID   | True    | 2622                  | 2397                   |
  | SAME    | False   | 691                   | 656                    |
  | SAME    | True    | 438                   | 396                    |

- I also tested the forward time only, but it's not a difference for me:
  ```py
  from tensorflow import keras

  def test_conv(use_fp16=True, padding="VALID"):
      if use_fp16:
          dtype = keras.mixed_precision.Policy('mixed_float16')
      else:
          dtype = "float32"
      inputs = tf.ones([512, 112, 112, 3])
      nn = keras.layers.Conv2D(64, 7, strides=2, use_bias=False, padding=padding, name="conv1_conv", dtype=dtype)
      print(nn(inputs).dtype)

      %timeit nn(inputs)

  test_conv(use_fp16=True, padding="VALID")  # 5.62 ms ± 82.2 µs
  test_conv(use_fp16=True, padding="SAME")   # 6.61 ms ± 109 µs
  test_conv(use_fp16=False, padding="VALID") # 4.77 ms ± 36.6 µs
  test_conv(use_fp16=False, padding="SAME")  # 5.14 ms ± 53.9 µs
  ```
- So what did I do wrong here?

**Related issue**
- I think this may related with issue [#39556](https://github.com/tensorflow/tensorflow/issues/39556) and [#41715](https://github.com/tensorflow/tensorflow/issues/41715). Here is my test results of `float32` / `float16` with `NHWC` / `NCHW`:
  ```py
  time for float32 NHWC: 0.11824584007263184
  time for float32 NCHW: 0.06493830680847168
  time for float16 NHWC: 0.1193842887878418
  time for float16 NCHW: 0.09624576568603516
  ```

**Comment by Saduf2019** [comment]
I am able to replicate this issue on yf 2.4 and nightly , please find the [gist here](https://colab.research.google.com/gist/Saduf2019/fc475839fb6be70958d587a665f676c4/untitled580.ipynb)

**Comment by leondgarse** [comment]
So does anyone checking this? Also tested `tf-nightly 2.6.0.dev20210401` with `python 3.8.5`, same situation.

**Comment by jvishnuvardhan** [comment]
@reedwm Any insights on the root-cause of the issue? Thanks!

**Comment by reedwm** [comment]
I cannot reproduce, either on TF 2.4 or the latest tf-nightly. I did use different GPUs (I tried both a Titan V and an A100), which could explain it.

I tried running the first program in the original post with the following commands on TF 2.4.1:

```python
python conv2d_slow.py --input_shape=112 --batch_size=512
python conv2d_slow.py --input_shape=112 --batch_size=512 --use_fp16
```

On a Titan V: In the former case (fp32), I get 54ms/step for the second epoch. In the latter case (fp16), I get 53ms/step.

@leondgarse, can you confirm I'm running the correct commands? Also, you mentioned you are using CUDA 11.2, but I believe TF 2.4 requires CUDA 11.0. `tf-nightly` currently requires CUDA 11.2 though (and has required it on the version you used), so that shouldn't have caused problems when you ran `tf-nightly`

**Comment by leondgarse** [comment]
Yes, yes, the commands are right, and I just noticed TF is actually loading `CUDA 11.0`, but my `nvidia-smi` is reporting `CUDA 11.2`. It's also `CUDA 11.0` I actually installed...
```sh
$ apt list --installed | grep cuda-cudart
cuda-cudart-dev-11-0/now 11.0.221-1 amd64 [installed,local]
```

I just noticed when the lag happens, the CPU usage is very low `<50%`, and GPU usage is very high `>90%`.
```py
CUDA_VISIBLE_DEVICES='1' python conv2d_slow.py -i 112 -b 512 -f
```
![Selection_340](https://user-images.githubusercontent.com/5744524/115483738-ee751c80-a283-11eb-8cf6-de83a9fc6a22.png)

When normal, CPU usage is high at `600~1000%`, and GPU `~20%`
```py
CUDA_VISIBLE_DEVICES='1' python conv2d_slow.py -i 112 -b 512
CUDA_VISIBLE_DEVICES='1' python conv2d_slow.py -i 112 -b 256 -f
```
![Selection_341](https://user-images.githubusercontent.com/5744524/115483883-398f2f80-a284-11eb-8efd-5e2b1f533dbf.png)

It's good to know this is not the common case, means I may find a solution. :)

**Comment by leondgarse** [comment]
Updated and tested using `tf-nightly 2.6.0-dev20210420` and `tensorflow 2.5.0-rc1` with `cuda 11.2.2-1` + `cudnn 8.1.1.33` + `nvidia-driver 465.19.01`. No luck, same situation...

**Comment by reedwm** [comment]
@nluehr can you take a look? I do not have access to any cards with compute capability 7.5

> Actually I used to have an issue running tf-nightly, Could not load dynamic library 'libcusolver.so.11'. I just made it work by sudo ln -s libcusolver.so libcusolver.so.11. Maybe my tf-nightly is not running correctly.

Unrelated to this issue, but the problem is you need CUDA 11.2. @sanjoy can we improve the error message if the wrong version of CUDA is used?

**Comment by leondgarse** [comment]
Tested with `pytorch==1.8.1+cu111`, and the result seems fine under the same environment:
- **Result**

  | input_shape | batch_size | use_fp16 | first epoch (ms/step) | second epoch (ms/step) |
  | ----------- | ---------- | -------- | --------------------- | ---------------------- |
  | 112         | 512        | False    | 84.5125               | 78.9935                |
  | 112         | 512        | True     | 75.3213               | 76.9712                |
  | 224         | 512        | False    | 228.2185              | 215.2186               |
  | 224         | 512        | True     | 228.9982              | 254.9154               |

- **Script**
  ```py
  import torch
  import time
  import torch.nn as nn
  import numpy as np
  from torchvision import datasets, transforms
  from tqdm import tqdm
  
  
  class TestResNet(nn.Module):
      def __init__(self, input=32, num_classes=10):
          super(TestResNet, self).__init__()
          self.conv1 = nn.Conv2d(3, 64, kernel_size=7, stride=2, padding=0, bias=False)
          self.flatten = nn.Flatten()
          conv_out_shape = int(np.ceil(input / 2)) - 3
          self.linear = nn.Linear(64 * conv_out_shape * conv_out_shape, num_classes)
          self.softmax = nn.Softmax(dim=-1)
  
      def forward(self, x):
          out = self.conv1(x)
          out = self.flatten(out)
          out = self.linear(out)
          out = self.softmax(out)
          return out
  
  
  def load_cifar10(batch_size=512, image_shape=(32, 32)):
      train_transforms = transforms.Compose([transforms.Resize(image_shape), transforms.ToTensor()])
      trainset = datasets.CIFAR10(root="./data", train=True, download=True, transform=train_transforms)
      train_loader = torch.utils.data.DataLoader(trainset, batch_size=batch_size, shuffle=True, num_workers=4)
      total = trainset.data.shape[0]
      steps_per_epoch = int(np.ceil(total / batch_size))
      return train_loader, steps_per_epoch
  
  
  def run_test(input_shape=(32, 32), batch_size=512, use_fp16=False, epochs=2):
      train_loader, steps_per_epoch = load_cifar10(batch_size=batch_size, image_shape=input_shape)
  
      device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
      model = TestResNet(input=input_shape[0], num_classes=10).to(device)
      if use_fp16:
          model = model.half()
  
      optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
      criterion = nn.CrossEntropyLoss()
      for epoch in range(epochs):
          since = time.time()
          for inputs, labels in tqdm(train_loader, "Epoch {}/{}".format(epoch + 1, epochs), total=steps_per_epoch):
              inputs = inputs.to(device).half() if use_fp16 else inputs.to(device)
              labels = labels.to(device)
              optimizer.zero_grad()
              outputs = model(inputs)
              loss = criterion(outputs, labels)
              loss.backward()
              optimizer.step()
  
          total_time = time.time() - since
          print(">>>> Total time: %.4fs, mean: %.4fms" % (total_time, total_time * 1000 / steps_per_epoch))
  
  
  if __name__ == "__main__":
      import sys
      import argparse
  
      parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
      parser.add_argument("-b", "--batch_size", type=int, default=512, help="Batch size")
      parser.add_argument("-i", "--input_shape", type=int, default=32, help="Input shape")
      parser.add_argument("-f", "--use_fp16", action="store_true", help="Use fp16")
      parser.add_argument("-e", "--epochs", type=int, default=2, help="Epochs")
  
      args = parser.parse_known_args(sys.argv[1:])[0]
      run_test((args.input_shape, args.input_shape), args.batch_size, args.use_fp16, args.epochs)
  ```

**Comment by leondgarse** [comment]
I just updated my `nvidia-driver` to `465.27`, as well as `cuda-driver==11.2.152-1` `libcudnn8==8.2.1.32-1`. This issue is gone on both my `Quadro RTX 8000` and `GeForce RTX 2080 Ti` now.

| input_shape | batch_size | use_fp16 | first epoch (ms/step) | second epoch (ms/step) |
| ----------- | ---------- | -------- | --------------------- | ---------------------- |
| 112         | 512        | False    | 71                    | 59                     |
| 112         | 512        | True     | 65                    | 54                     |
